S3Box code review

cgclever opened the pull request:
**Overview**

S3Box is a manager for easily and uniformly streaming data to multiple files, after which the user can create Redshift manifest files pointing to said files.

Here I've siphoned off a new directory meant explicitly for external code review.

**Open Questions**

1. Should `Pack` be more lenient about checking for JSON inputs? Arguably S3Box should be agnostic and more flexible, while a Redbox will want to create the checks.
2. Should we have a `Seal` method, or should that happen behind the scenes in the `CreateManifests` function? One disadvantage of `Seal` is it's another step for the user. One advantage of `Seal` is that it prevents writes and alerts the user that no more writes should be occurring.

cgclever commented on s3box_review/s3box.go:
```diff
@@ -0,0 +1,243 @@
+	oldBuffer := sb.bufferedData // If write fails, keep buffered data unchanged
+	data = append(data, '\n')    // Append a new line for text-editor readability
+	sb.bufferedData = append(sb.bufferedData, data...)
+	sb.Unlock()
```
I'm thinking of changing this to
```
defer sb.Unlock()
```
and keep the entire pack locked to one at a time.

cgclever commented on s3box_review/s3box.go:
```diff
@@ -0,0 +1,243 @@
+		return nil
+	}
+	fileKey := fmt.Sprintf("%d_%d.json.gz", sb.timestamp.UnixNano(), sb.fileNumber)
+	sb.Lock()
```
Thinking of moving the lock above. If `Seal` is called twice in quick succession, the first call will upload the remaining data, while the second call will attempt to write nothing. This wouldn't be disastrous, but would add noise and messiness to s3.

This being said, the user should _never_ need to `Seal` more than once.

bstein-clever replied:
Shouldn't? No.
Might anyway? Who knows?

I'm a big fan of protecting users against their own stupidity and failing to read docs.
Even returning an error ("You already called Seal, don't do that!") seems better than doing something undesirable.

cgclever replied:
Yep, pushed the change to put the Lock/Unlock in a better place.

kvigen commented on s3box_review/README.md:
```diff
@@ -0,0 +1,126 @@
+
+### CreateManifests
+
+`func CreateManifests(manifestKey string, numManifests int) ([]string, error)`
```
is there a reason a user would want to seal, but not create a manifest? i'm wondering whether it makes sense to combine these two functions?

cgclever replied:
I've been wondering the same. The verbiage was intuitive to me that you would "Pack and Seal". The primary value of "Seal" would be to protect yourself to ensure no more writes can happen.

kvigen commented on s3box_review/README.md:
```diff
@@ -0,0 +1,126 @@
+
+### NextBox
+
+`func NextBox()`
```
i didn't find this name quite as intuitive as the others. i've seen `Clone()` used for similar things, but I don't think that's quite right either

cgclever replied:
Maybe `NewBox`? `FreshBox`?

kvigen replied:
i like `NewBox`

schimmy replied:
hmm, seems like the docs need to be regenerated, as there's no NextBox anymore!

kvigen commented on s3box_review/s3_helpers.go:
```diff
@@ -0,0 +1,102 @@
+func getRegionForBucketProd(name string) (string, error) {
+	// Any region will work for the region lookup, but the request MUST use PathStyle
+	config := aws.NewConfig().WithRegion("us-west-1").WithS3ForcePathStyle(true)
+	session := session.New()
```
do we have to close sessions? i'm not sure, just asking

cgclever replied:
This is a great question. The logic is currently written so that only one s3 call should ever happen at once. However if we opened this up in the future, we'd probably want to clone the sessions. FWIW, I've not seen a closed AWS session in Clever code, however `Copy` is an available method.

kvigen commented on s3box_review/s3box.go:
```diff
@@ -0,0 +1,243 @@
+
+const (
+	// DefaultBufferSize is set to 10MB
+	DefaultBufferSize = 10000000
```
nit: make this `10 * 1000 * 1000` for readability

kvigen commented on s3box_review/s3box.go:
```diff
@@ -0,0 +1,243 @@
+	timestamp time.Time
+
+	// fileNumber indicates the number of s3 files which have currently been created
+	fileNumber int
```
is this redundant with `len(fileLocations)`?

cgclever replied:
It is, just felt it was nicer not to always have to write `len(fileLocations)`. However there's no issue in removing this redundancy.

schimmy replied:
I think it's more go-like to remove, and less likely to get us into trouble

kvigen commented on s3box_review/s3box.go:
```diff
@@ -0,0 +1,243 @@
+
+	// If AWS creds were provided use those, otherwise grab them from your environment
+	var awsCreds *credentials.Credentials
+	if options.AWSKey == "" && options.AWSPassword == "" && options.AWSToken == "" {
```
what happens if only some of these are specified?

cgclever replied:
I've thought about it and feels somewhat of an edge case. If you fill in just one, then you'll simply get unresponsive s3 calls.

How would you approach this? Do an "or" instead?

kvigen replied:
too bad Go doesn't have ternaries :)

cgclever replied:
I feel like that behavior would be odd, because you could end up overriding some credentials and not others. Overall it feels unusual that someone would pass, say just an AWSKey or just an AWSPassword. If they try doing that, then the job will error and the user would figure out sooner than later what's going on.

I'm open to suggestions here in general, but my preference would be for there to be some type of erroring if it's configured incorrectly. My only thought would be a second check like 

```
else {
  if AWSKey == "" || AWSPassword == "" {
    return nil, fmt.Errorf("Error, missing key or Password")
  }
}
```

cgclever replied:
I went ahead and did this

kvigen commented on s3box_review/s3box.go:
```diff
@@ -0,0 +1,243 @@
+}
+
+// NextBox gives you a new box, forgetting everything about previously packaged data
+func (sb *S3Box) NextBox() {
```
random idea: should this return a new box instead of clearing out the old one?

cgclever replied:
We could do that. The reason I didn't before is because it would be recreating the s3 connection which feels like a little overhead. Whether or not it's nice, this way also has the property that "NextBox" won't error.

What's your preference here?

cgclever replied:
I'm going to get rid of it until a clear use case is needed. In order to go along with this suggestion we'd be returning a new box for the user to use, rather than clearing out the existing one.

kvigen commented on s3box_review/README.md:
```diff
@@ -0,0 +1,126 @@
+    handleError(sb.Pack(rowBytes))
+  }
+
+  // Manifest creation and data transfer to Redshift
```
do we need to add a seal in this example?

cgclever replied:
Yes we do :)

kvigen commented on s3box_review/README.md:
```diff
@@ -0,0 +1,126 @@
+    rowBytes, _ := json.Marshal(rowData)
+    handleError(sb.Pack(rowBytes))
+  }
+
```
i can't remember if we already discussed this, but should we add an abort method? i guess all it gives us is cleaning up s3 files... not sure how important that is

cgclever replied:
I wondered the same thing. Cleaning up the files would be nice, but doesn't feel crucial. However might be nice for completeness sake.

schimmy replied:
I agree with Christian - would say it's not a blocker for us using this, but would add it for yourself as a github "issue" so that you don't forget the idea

kvigen replied:
👍

kvigen commented on s3box_review/s3_helpers.go:
```diff
@@ -0,0 +1,102 @@
+	var wg sync.WaitGroup
+	var writeErr error
+	var streamErr error
+	wg.Add(2)
```
not a must-do, but might be worth looking at https://godoc.org/github.com/facebookgo/errgroup which combines wait groups and errors nicely

kvigen commented on s3box_review/s3box.go:
```diff
@@ -0,0 +1,243 @@
+
+// Pack writes bytes into a buffer. Once that buffer hits capacity, the data is output to s3.
+// Any error will leave the buffer unmodified.
+func (sb *S3Box) Pack(data []byte) error {
```
or not validate that it's json and leave that to redbox, or whatever layer someone wants to build on top of this

cgclever replied:
I like leaving any validation checks to `Redbox` an letting `S3Box` be a simple streaming manager to s3.

schimmy replied:
Curious to see how this interface question plays out in practice!  I guess making it a []byte does give slightly more flexibility...

cgclever commented on s3box_review/s3_helpers.go:
```diff
@@ -0,0 +1,102 @@
+}
+
+func init() {
+	getRegionForBucket = getRegionForBucketProd
```
The reason I did this was to make testing more manageable.

The primary alternative I saw was to have an `S3Box` take in some input like `S3Manager` which would default to these, but I worried it would overcomplicate the initialization.

schimmy replied:
there's something here that still triggers a spidey sense of some kind, but can't fully place it... but I'm not too worried, I think this will be fine too

cgclever commented on s3box_review/s3box.go:
```diff
@@ -0,0 +1,243 @@
+	// If the bytes aren't in JSON format, return an error
+	var tempMap map[string]interface{}
+	if err := json.Unmarshal(data, &tempMap); err != nil {
+		return ErrInvalidJSONInput
```
I primarily did this to make testing super thorough. But I'm going to follow Kyle's suggestion and remove the check in `S3Box`

cgclever commented:
@schimmy: Addressed most PR comments.

Major decisions:
- Remove Seal function
- Remove NextBox function
- Pack doesn't check for JSON input

Minor changes:
- Adjust where locks go
- Remove fileNumber field
- Add AWS creds check
